src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs

```diff
@@ -3001,6 +3010,35 @@ private BoundExpression BindStackAllocArrayCreationExpression(StackAllocArrayCre
             return new BoundStackAllocArrayCreation(node, count, pointerType, hasErrors || typeHasErrors);
         }
 
+        private bool IsLegalStackAllocLocation(StackAllocArrayCreationExpressionSyntax node)
+        {
+            if (node.Parent == null ||
+                node.Parent.Kind() != SyntaxKind.EqualsValueClause ||
+                node.Parent.Parent.Kind() != SyntaxKind.VariableDeclarator ||
+                node.Parent.Parent.Parent.Kind() != SyntaxKind.VariableDeclaration)
+            {
+                return false;
+            }
+
+            var variableDeclaration = (VariableDeclarationSyntax)node.Parent.Parent.Parent;
+            if (variableDeclaration.Parent.Kind() == SyntaxKind.FieldDeclaration ||
+                variableDeclaration.Parent.Kind() == SyntaxKind.EventFieldDeclaration)
+            {
+                return false;
+            }
+
+            if (variableDeclaration.Parent.Kind() == SyntaxKind.LocalDeclarationStatement)
+            {
+                var localDeclaration = (LocalDeclarationStatementSyntax)variableDeclaration.Parent;
+                if (localDeclaration.Modifiers.Any(SyntaxKind.ConstKeyword))
+                {
+                    return false;
+                }
+            }
```

Comment:
If not changing to `switch`, please extract a local for `variableDeclaration.Parent` or a `parent` local updated repeatedly in the first `if` as suggested by @jcouv.